generate url from the selected mirror in apt-gen-list

OriginCode opened the pull request.

OriginCode committed:
generate url from the selected mirror in apt-gen-list

liushuyu commented on Cargo.toml:
```diff
@@ -11,6 +11,8 @@ anyhow = "1.0"
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
+serde_yaml = "0.8"
+lazy_static = "1.4.0"
```
You don't need to pin the version of this crate:

```suggestion
lazy_static = "1.4"
```

liushuyu commented on src/pm.rs:
```diff
@@ -1,21 +1,58 @@
 const DPKG_STATE: &str = "/var/lib/dpkg/status";
-const MIRROR_URL: &str = "https://repo.aosc.io/debs";
+
+fn get_mirror_url() -> Result<String> {
+    let mut status_file = fs::File::open("/var/lib/apt/gen/status.json")?;
+    let mut status_data = String::new();
+    status_file.read_to_string(&mut status_data)?;
+
+    let status_v: jsonValue = serde_json::from_str(&status_data)?;
+    let mirror: String = match &status_v["mirror"][0] {
+        jsonValue::String(m) => String::from(m),
+        _ => {
+            return Err(format_err!("Malformed apt-gen-list data."));
+        }
+    };
+
+    let mut mirrors_file = fs::File::open("/usr/share/distro-repository-data/mirrors.yml")?;
+    let mut mirrors_data = String::new();
+    mirrors_file.read_to_string(&mut mirrors_data)?;
+
+    let mirrors_v: yamlValue = serde_yaml::from_str(&mirrors_data)?;
+    let mirror_url = match &mirrors_v[&mirror]["url"] {
+        yamlValue::String(u) => String::from(u),
+        _ => {
+            return Err(format_err!("Malformed mirrors data."));
+        }
+    };
+
+    Ok(mirror_url)
+}
+
+lazy_static! {
+    static ref MIRROR_URL: String = format!(
+        "{}{}",
+        get_mirror_url().unwrap_or(String::from("https://repo.aosc.io/")),
+        "debs"
+    );
+}
```
Is this function the same as that in https://github.com/AOSC-Dev/atm/pull/1/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR80-R115?

Why is this duplicated?

OriginCode replied:
Thus do I need to write a separate file to be imported by these two files?

liushuyu replied:
> Thus do I need to write a separate file to be imported by these two files?

No. You need to mark your function as either `pub` or `pub(crate)` and then you can import it.

OriginCode replied:
Do I need to create a new `mod` or simply mark a `pub` ?

liushuyu replied:
You don't need a new module. Just mark the function as public.

liushuyu commented on src/main.rs:
```diff
@@ -1,18 +1,22 @@
 use std::time::Duration;
 use std::{cmp::Ordering, collections::HashSet};
 
+use anyhow::{ format_err, Result };
```
Remove the whitespaces around the brackets:

```suggestion
use anyhow::{format_err, Result};
```

liushuyu commented on src/main.rs:
```diff
@@ -73,6 +77,42 @@ macro_rules! unwrap_or_show_error {
+        }
+    };
+
+    let mut mirrors_file = File::open("/usr/share/distro-repository-data/mirrors.yml")?;
+    let mut mirrors_data = String::new();
+    mirrors_file.read_to_string(&mut mirrors_data)?;
```
There's a more simple way to read the file that's not very large in size.

liushuyu commented on src/main.rs:
```diff
@@ -73,6 +77,42 @@ macro_rules! unwrap_or_show_error {
+    let mut status_data = String::new();
+    status_file.read_to_string(&mut status_data)?;
+
+    let status_v: jsonValue = serde_json::from_str(&status_data)?;
+    let mirror: String = match &status_v["mirror"][0] {
+        jsonValue::String(m) => String::from(m),
+        _ => {
+            return Err(format_err!("Malformed apt-gen-list data."));
+        }
+    };
```
There is a better way to do this deserialize song and dance, take a look in https://github.com/AOSC-Dev/atm/blob/master/src/network.rs?

liushuyu commented on src/main.rs:
```diff
@@ -73,6 +77,42 @@ macro_rules! unwrap_or_show_error {
+    Ok(mirror_url)
+}
+
+lazy_static! {
+    static ref DEFAULT_MANIFEST_URL: String = format!(
+        "{}{}",
+        get_mirror_url().unwrap_or(String::from("https://repo.aosc.io/")),
+        "debs/manifest/topics.json"
+    );
+}
```
Hmm, before digging into this, let me ask you a question: How many times are `DEFAULT_MANIFEST_URL` referenced?

OriginCode replied:
Once?

liushuyu replied:
> Once?

So in that case, do you really need a global reference?

OriginCode replied:
So I can just write `format!` in that scope?

liushuyu replied:
Correct.

liushuyu reviewed (changes requested):
Overall, it's not too bad for a first-time contributor.

However, there are many soundness issues in the changes. You can run `cargo clippy` to see if there are other issues lying around. Also, before submitting changes, run `cargo-fmt` to format the code.

OriginCode committed:
unpin the version of lazy_static

OriginCode committed:
optimize the function:
1. move the function to pm.rs and import it in main.rs to prevent
   duplication
2. use serde deserialization instead of using Value from serde_json and
   serde_yaml

liushuyu commented on src/pm.rs:
```diff
@@ -15,7 +16,53 @@ const SOURCE_PATH: &str = "/etc/apt/sources.list.d/atm.list";
+    Ok(mirror_url)
+}
+
+lazy_static! {
+    static ref MIRROR_URL: String = format!(
+        "{}{}",
+        get_mirror_url().unwrap_or(String::from("https://repo.aosc.io/")),
+        "debs"
+    );
+}
```
Is this still needed?

OriginCode replied:
There's a `format!` at where `MIRROR_URL` is used. If directly use the `format!` inside that `format!` it would raise `doesn't have a size known at compile-time` err.

liushuyu replied:
However, you have the option to integrate this `format!` into the other one by flattening the formatting string.

liushuyu replied:
On the second thought, now that you have duplicate `format!`'s in `main` and `pm`, you could consider exporting `MIRROR_URL`, so it will be run only once.

liushuyu commented on src/pm.rs:
```diff
@@ -15,7 +16,53 @@ const SOURCE_PATH: &str = "/etc/apt/sources.list.d/atm.list";
+}
+
+pub fn get_mirror_url() -> Result<String> {
+    let mut status_file = fs::File::open("/var/lib/apt/gen/status.json")?;
```
> "/var/lib/apt/gen/status.json"

I recommend extracting these constant literal values to `const` variables.

liushuyu commented on src/pm.rs:
```diff
@@ -15,7 +16,53 @@ const SOURCE_PATH: &str = "/etc/apt/sources.list.d/atm.list";
+    status_file.read_to_string(&mut status_data)?;
+
+    let status_v: AptGenListStatus = serde_json::from_str(&status_data)?;
+    let mut mirror = String::from("origin");
+    if let Some(m) = &status_v.mirror.get(0) {
+        mirror = m.to_string();
+    }
```
You can actually do this:

```suggestion
    let mirror;
    if let Some(m) = &status_v.mirror.get(0) {
        mirror = m.to_string();
    } else {
        mirror = "origin".to_string();
    }
```

... to avoid using `mut`.

OriginCode committed:
optimize the gen_mirror_url function:
1. export MIRROR_URL variable for main.rs to use to prevent running the
   function twice
2. extract constant file paths to const variables

OriginCode committed:
pm: format code
